### PR TITLE
Result marginal counts

### PR DESCRIPTION
**Pull request by ajavadia**
A utility function `qiskit.result.marginal_counts()` to obtain marginalized counts from an experiment where more measurements were done than desired.
I'll later deprecate the method that exists in ignis tomography module (https://github.com/Qiskit/qiskit-ignis/issues/283)

### CODE REVIEW

File: qiskit/result/utils.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def marginal_counts(counts, indices=None, pad_zeros=False):
+    """Marginalize counts from an experiment over some indices of interest.
+
+    Args:
+        counts (dict): Counts of some measurement results, to be marginalized.
+        indices (list(int) or None): The bit positions of interest to NOT marinalize over.
+            If None, do not marginalize at all.
+        pad_zeros (Bool): Include zero count outcomes in return dict.
```
Comment (taalexander) [inline qiskit/result/utils.py]:
```suggestion
        pad_zeros (bool): Include zero count outcomes in return dict.
```

File: qiskit/result/utils.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    Returns:
+        dict[str:int]: a dictionary with the observed counts, marginalized to
```
Comment (taalexander) [inline qiskit/result/utils.py]:
```suggestion
        Dict[str, int]: a dictionary with the observed counts, marginalized to
```

File: qiskit/result/utils.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+from re import match
+from qiskit.exceptions import QiskitError
+
+
+def marginal_counts(counts, indices=None, pad_zeros=False):
+    """Marginalize counts from an experiment over some indices of interest.
+
+    Args:
+        counts (dict): Counts of some measurement results, to be marginalized.
+        indices (list(int) or None): The bit positions of interest to NOT marinalize over.
```
Comment (taalexander) [inline qiskit/result/utils.py]:
```suggestion
        indices (Optional[List[int]]): The bit positions of interest to NOT marinalize over.
```

File: qiskit/result/utils.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+from functools import reduce
+from re import match
+from qiskit.exceptions import QiskitError
+
+
+def marginal_counts(counts, indices=None, pad_zeros=False):
+    """Marginalize counts from an experiment over some indices of interest.
+
+    Args:
+        counts (dict): Counts of some measurement results, to be marginalized.
```
Comment (taalexander) [inline qiskit/result/utils.py]:
```suggestion
        counts (Dict): Counts of some measurement results, to be marginalized.
```

Comment (ajavadia) [reply qiskit/result/utils.py]: this is following the rest of docstrings in qiskit, which is not quite the same as the pulse package.

**Comment by ajavadia** [comment]
The raw results are returned in hex. The `get_counts` function does some postprocessing for human consumption, including e.g. inserting spaces between classical registers (`100 01`). The thinking is that for small devices it's easier for the user to interpret bitstrings rather than integers or hex.

**Comment by taalexander** [comment]
I do agree with that, I guess what I'm saying is that if the `Result` was an object instead of a dictionary we could have all of those representations in one with different accessor methods. The downside would be serialization, but that could always be accessible with a method like `as_dict`.

**Comment by ajavadia** [comment]
> if the Result was an object instead of a dictionary

That is the case already.

Also this PR should be ready now.

**Comment by taalexander** [comment]
Yes and all I'm saying is we could store internally as an integer which should make the manipulations such as marginal_counts easier as they would be done with bit-banging. Results could still be output with accessor methods like they are currently in the chosen format.

**Pull request merged**